src/model/mod.rs

```diff
@@ -48,8 +50,145 @@ pub enum PlayingItem {
     Episode(show::FullEpisode),
 }
 
+#[derive(Debug, PartialEq, Eq, Clone)]
+pub struct Id<'id> {
+    _type: Type,
+    id: &'id str,
+}
+
+#[derive(Debug, PartialEq, Eq, Clone, Copy, Display, Error)]
+pub enum IdError {
+    InvalidPrefix,
+    InvalidFormat,
+    InvalidType,
+    InvalidId,
+}
+
+impl std::fmt::Display for Id<'_> {
+    fn fmt(&self, f: &mut std::fmt::Formatter) -> std::fmt::Result {
+        f.write_str("spotify:")?;
+        f.write_str(self._type.as_ref())?;
+        f.write_str(":")?;
+        f.write_str(self.id)?;
+        Ok(())
+    }
+}
+
+impl AsRef<str> for Id<'_> {
+    fn as_ref(&self) -> &str {
+        self.id
+    }
+}
+
+impl Id<'_> {
+    pub fn _type(&self) -> Type {
+        self._type
+    }
+
+    pub fn id(&self) -> &str {
+        &self.id
+    }
+
+    pub fn uri(&self) -> String {
+        format!("spotify:{}:{}", self._type.as_ref(), self.id)
+    }
+
+    pub fn from_id_or_uri<'a, 'b: 'a>(_type: Type, id_or_uri: &'b str) -> Result<Id<'a>, IdError> {
+        match Self::from_uri(id_or_uri) {
+            Ok(id) if id._type == _type => Ok(id),
+            Ok(_) => Err(IdError::InvalidType),
+            Err(IdError::InvalidPrefix) => Self::from_id(_type, id_or_uri),
+            Err(error) => Err(error),
+        }
+    }
+
+    pub fn from_id<'a, 'b: 'a>(_type: Type, id: &'b str) -> Result<Id<'a>, IdError> {
+        if id.chars().all(|ch| ch.is_ascii_alphanumeric()) {
+            Ok(Id { _type, id })
+        } else {
+            Err(IdError::InvalidId)
+        }
+    }
+
+    pub fn from_uri<'a, 'b: 'a>(uri: &'b str) -> Result<Id<'a>, IdError> {
+        if let Some((tpe, id)) = if let Some(uri) = uri.strip_prefix("spotify:") {
```

Comment:
Having nested `if let`s makes the code much harder to read. Is it possible to break this up a bit? For instance we could:

1. Strip the starting `spotify`
2. Compare the first character. If it's not `:` or `/`, return an error.
3. Use that character to split the remaining slice into what should be:
    1. An empty slice
    2. The type, which can be parsed
    3. The ID, which can be passed to `from_id`

Reply:
I don't like having an empty slice. It feels like an unnecessary splitting job.